File: onyx-database/src/com/onyx/descriptor/IndexDescriptor.java
package com.onyx.descriptor;

/**
 * Created by timothy.osborn on 12/11/14.
 */
public class IndexDescriptor extends AbstractBaseDescriptor implements BaseDescriptor {

    protected String type = null;
    protected EntityDescriptor entityDescriptor;

    public EntityDescriptor getEntityDescriptor() {
        return entityDescriptor;
    }

    public void setEntityDescriptor(EntityDescriptor entityDescriptor) {
        this.entityDescriptor = entityDescriptor;
    }

    protected byte loadFactor = 1;

    /**
     * This method is to determine what scale the underlying structure should be.  The values are from 1-10.
     * 1 is the fastest for small data sets.  10 is to span huge data sets intended that the performance of the index
     * does not degrade over time.  Note: You can not change this ad-hoc.  You must re-build the index if you intend
     * to change.  Always plan for scale when designing your data model.
     *
     * @param loadFactor Value from 1-10.
     * @since 1.2.0
     */
    public void setLoadFactor(byte loadFactor) {
        this.loadFactor = loadFactor;
    }

    /**
     * Getter for the load factor
     *
     * @return The values are from 1-10.
     * <p>
     * 1 is the fastest for small data sets.  10 is to span huge data sets intended that the performance of the index
     * does not degrade over time.  Note: You can not change this ad-hoc.  You must re-build the index if you intend
     * to change.  Always plan for scale when designing your data model.
     * @since 1.2.0
     */
    public byte getLoadFactor() {
        return this.loadFactor;
    }

    ////////////////////////////////////////////////////////
    //
    //  Hashing Object Overrides
    //
    ////////////////////////////////////////////////////////

    /**
     * Used for calculating hash structure
     *
     * @return
     */
    @Override
    public int hashCode() {
        if (entityDescriptor.getPartition() != null) {
            (getType().getName() + getName() + entityDescriptor.getPartition().getPartitionValue()).hashCode();
        }
        return (getType().getName() + getName() + entityDescriptor.getClazz().getName()).hashCode();
    }

    /**
     * Used for usage within a hashmap
     *
     * @param val
     * @return
     */
    @Override
    public boolean equals(Object val) {
        if (val instanceof IndexDescriptor) {
            IndexDescriptor comparison = (IndexDescriptor) val;

            if (((IndexDescriptor) val).loadFactor != this.loadFactor)
                return false;

            if (!entityDescriptor.clazz.getName().equals(comparison.entityDescriptor.clazz.getName())) {
                return false;
            }

            if (entityDescriptor.getPartition() != null && comparison.getEntityDescriptor().getPartition() != null) {
                return (comparison.getType().getName().equals(getType().getName())
                        && comparison.getName().equals(getName())
                        && comparison.getEntityDescriptor().getPartition().getPartitionValue().equals(entityDescriptor.getPartition().getPartitionValue()));
            } else if (entityDescriptor.getPartition() != null) {
                return false;
            } else if (comparison.getEntityDescriptor().getPartition() != null) {
                return false;
            }


            return (((IndexDescriptor) val).getType().getName().equals(getType().getName())
                    && ((IndexDescriptor) val).getName().equals(getName()));
        }
        else if(val instanceof IdentifierDescriptor)
        {
            return this.loadFactor == ((IdentifierDescriptor) val).loadFactor
                    && this.name.equals(((IdentifierDescriptor) val).name);
        }

        return false;
    }
}

File: onyx-database/src/main/kotlin/com/onyx/descriptor/IndexDescriptor.kt
package com.onyx.descriptor

import com.onyx.extension.common.ClassMetadata
import com.onyx.persistence.annotations.values.IndexType

/**
 * Created by timothy.osborn on 12/11/14.
 *
 * General information regarding an index within an entity
 */
open class IndexDescriptor(
    override var name: String = "",
    open var type: Class<*> = ClassMetadata.ANY_CLASS,
    open var indexType: IndexType = IndexType.DEFAULT,
    open var entropy: Int = 0,
    open var encodingVersion: Int = 0,
    open var configurationId: Long = 0L,
    open var configurationSignature: String = ""
) : AbstractBaseDescriptor(), BaseDescriptor {

    open lateinit var entityDescriptor: EntityDescriptor

    override fun hashCode(): Int {
        var result = (((System.identityHashCode(this.entityDescriptor.entityClass)) * 31) * 31 + this.name.hashCode()) * 31
        result = 31 * result + entityDescriptor.partition.hashCode()
        return result
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is IndexDescriptor) return false

        // Compare partition using equals (which now uses identity for Class)
        if (this.entityDescriptor.partition != other.entityDescriptor.partition) return false
        if (this.name != other.name) return false

        return true
    }
}
